share/getters/ipld.go

```diff
@@ -0,0 +1,139 @@
+func NewIPLDGetter(bServ blockservice.BlockService) *IPLDGetter {
+	return &IPLDGetter{
+		rtrv:  eds.NewRetriever(bServ),
+		bServ: bServ,
+	}
+}
+
+func (ig *IPLDGetter) GetShare(ctx context.Context, dah *share.Root, row, col int) (share.Share, error) {
+	root, leaf := ipld.Translate(dah, row, col)
+	blockGetter := getGetter(ctx, ig.bServ)
+	nd, err := share.GetShare(ctx, blockGetter, root, leaf, len(dah.RowsRoots))
+	if err != nil {
+		return nil, fmt.Errorf("getter/ipld: failed to retrieve share: %w", err)
+	}
+
+	return nd, nil
+}
+
+func (ig *IPLDGetter) GetEDS(ctx context.Context, root *share.Root) (*rsmt2d.ExtendedDataSquare, error) {
+	// rtrv.Retrieve calls shares.GetShares until enough shares are retrieved to reconstruct the EDS
+	eds, err := ig.rtrv.Retrieve(ctx, root)
+	if err != nil {
+		return nil, fmt.Errorf("getter/ipld: failed to retrieve eds: %w", err)
+	}
+	return eds, nil
+}
+
+func (ig *IPLDGetter) GetSharesByNamespace(
+	ctx context.Context,
+	root *share.Root,
+	nID namespace.ID,
+) (share.NamespaceShares, error) {
+	if len(nID) != share.NamespaceSize {
+		return nil, fmt.Errorf("getter/ipld: expected namespace ID of size %d, got %d",
+			share.NamespaceSize, len(nID))
+	}
+
+	rowRootCIDs := make([]cid.Cid, 0, len(root.RowsRoots))
+	for _, row := range root.RowsRoots {
+		if !nID.Less(nmt.MinNamespace(row, nID.Size())) && nID.LessOrEqual(nmt.MaxNamespace(row, nID.Size())) {
+			rowRootCIDs = append(rowRootCIDs, ipld.MustCidFromNamespacedSha256(row))
+		}
+	}
+	if len(rowRootCIDs) == 0 {
+		return nil, nil
+	}
+
+	blockGetter := getGetter(ctx, ig.bServ)
+	errGroup, ctx := errgroup.WithContext(ctx)
+	shares := make([]share.RowNamespaceShares, len(rowRootCIDs))
+	for i, rootCID := range rowRootCIDs {
+		// shadow loop variables, to ensure correct values are captured
+		i, rootCID := i, rootCID
+		errGroup.Go(func() error {
+			proof := new(ipld.Proof)
+			row, err := share.GetSharesByNamespace(ctx, blockGetter, rootCID, nID, len(root.RowsRoots), proof)
+			shares[i] = share.RowNamespaceShares{
+				Shares: row,
+				Proof:  proof,
+			}
+			if err != nil {
+				return fmt.Errorf("getter/ipld: retrieving nID %x for row %x: %w", nID, rootCID, err)
+			}
+			return nil
+		})
+	}
+
+	if err := errGroup.Wait(); err != nil {
+		return nil, err
+	}
+
+	return shares, nil
+}
+
+// ses is a struct that can optionally be passed by context to the share.Getter methods using
+// WithSession to indicate that a blockservice session should be created.
+type ses struct {
+	sync.Mutex
+	atomic.Pointer[blockservice.Session]
+}
+
+// WithSession stores an empty ses in the context, indicating that a blockservice session should be
+// created.
+func WithSession(ctx context.Context) context.Context {
+	return context.WithValue(ctx, ses{}, &ses{})
+}
+
+func getGetter(ctx context.Context, service blockservice.BlockService) blockservice.BlockGetter {
+	s, ok := ctx.Value(ses{}).(*ses)
```

Comment:
I just realized that there is actually a bug here. The key should be the same value. The go doc on Value method recommends making private global var used as a key.

Reply:
use an empty struct now for the key

Reply:
I think the suggestion by godoc is to use same global `var` (e.g. value of a type) rather than creating new value every time. Also it could be empty struct of `ses` since you can use any comparable type as a key. Struct is comparable and empty struct == another empty struct.

Reply:
Yeah, we don't need a new type there, but a same value which can be the same session type

Reply:
Found some interesting things inspecting bytecode that I thought I would share:

 1. creating a new type of `type sessionKey struct{}` and then using `sessionKey{}` has the same bytecode as
 ```golang
 type key struct{}
 var sessionKey = key{}
 ``` 
 and using `sessionKey`. I initially picked this new type because I compared it to `var sessionKey = session{}` with `sessionKey` as the key, which produces a lot more bytecode (it copies the value, and linter complains), BUT i had forgotten to check:
 
 2. using `var sessionKey = session{}` and then using `&sessionKey` as the key has the same bytecode as above (except with different addresses ofc)
 
going for #2 as you've suggested

Reply:
First approach has more bytecode because to compare structs golang compares struct fields. An if struct filed happens to be a struct it does the same recursively (for mutex fields inside session). Comparing pointer on the other hand is just int64 compare.

Reply:
> Comparing pointer on the other hand is just int64 compare

Wait, but if we take two different ptrs from the same value, those pointers would be different

Reply:
cc @distractedm1nd